doc/source/configuration.rst
********************************
Setting Up and Configuring a Run
********************************

.. highlight:: xml

Introduction
============

All Ginkgo simulations are driven by an XML-format configuration file.
As with any valid XML document, the configuration file should begin with an XML version declaration (i.e, ``<?xml version="1.0"?>``).
Following this, the configuration file should have a ``<ginkgo>`` element as its root level element.
This element has various attributes that customize logging, reporting and other output of the simulation, and are described below.
The ``<ginkgo>`` element has the following top-level elements which define in the detail the actual simulation and the sampling of genealogies:

    ``<system>``
        This defines details of the simulation framework, such as the the number of dimensions to the fitness equation, the number of generations for which to run the simulation, the global selection strength, etc.

    ``<landscape>``
        This defines the dimensions of the landscape of the simulation.

    ``<lineages>``
        This defines the different species ecologies of the organisms that populate and interact in the landscape.

    ``<initialization>``
        This defines the "burn-in" or "bootstrapping" phase of the simulation, where the landscape is populated and life-cycles are run until the population structures of the organsms are conditioned under the specified environment.

    ``<environments>``
        This defines the changing geographic and environmental templates of the landscapes at different points in the simulation.

    ``<samples>``
        This defines the schedule of genealogies and occurrence samples that takes place during the course of the simulation.

Thus, a typical configuration file would have a structure line this::

    <?xml version="1.0"?>
    <ginkgo>
        <system> ... </system>
        <landscape> ... </landscape>
        <lineages> ... </lineages>
        <initialization> ... </initialization>
        <environments> ... </environments>
        <samples> ... </samples>
    </ginkgo>

Configuration Elements
======================


The ``<ginkgo>`` Element: Defining the Reporting Parameters
-----------------------------------------------------------

This is the (mandatory) root element of the configuration document, and contains as its children all other elements that, taken together, provide that directives that define and drive the simulation, from start to initialization to the main cycles to reporting and output.

This ``<ginkgo>`` can optionally have the following attributes:

    ``title``
        This defines a common prefix for all output files from a run of this simulation. If the "``-i``" command-line options is passed to the Ginkgo invocation, the identifier given by the "``-i``" flag will be appended to the end of the string given by this attribute. Typically, the ``title`` attribute would be used to provide some sort of description of the scenario being simulated, while the command-line "``-i``" flag describes the replicate number or identity.
        This combination of an optional replicate identifier in the Ginkgo invocation and simulation schema ttile allow for the running multiple realizations of the same simulation regime, but maintain distinct sets of output files for each realization in the same directory, as each output file will have the prefix "``<title><i>``".

    ``log_frequency``
        A non-zero positive integer specifying frequency of log output in terms of numbers of generations. Defaults to 10.

    ``multifurcating_trees``
        If "``True``", then multifurcating trees are preserved. If "``False``", then multifurcating trees are converted into bifurcating trees by arbitrarily resolving the polytomies with new edges of length 0. Defaults to "``True``" (i.e., preserve polytomies).

    ``final_output``
        By default, at the end of the simulation Ginkgo samples and reports the genealogies, traits and occurrences of *every* organism of *every* lineage in *every* cell.  Setting ``final_output`` to "``False``" suppresses the automatic reporting of the full results, thus restricting the output to a own custom fine-tuned sampling regime. Note that if ``final_output`` is set to "``False``", if no ``<sample>`` directives are given, then no output will be produced apart from the log files!

    ``full_complement_diploid_trees``
        By default, the genealogies for the diploid locii are built by randomly sampling one of the alleles of the diploid complement for each organism (and saved with the "``.diploid1.tre``" extension). By setting this attribute to "``False``", you can ask that Ginkgo produce a file (which will be saved with the "``.diploid2.tre``"  extension) where *both* alleles are sampled.

For example::

    <ginkgo
        title="results"
        log_frequency="1"
        multifurcating_trees="True"
        final_output = "True"
        full_complement_diploid_trees = "False">
        :
        :
    </ginkgo>


The ``<system>`` Element: Defining the Simulation Meta-Parameters
-----------------------------------------------------------------

The mandatory ``<system>`` element defines some meta-level aspects of the simulation in its child elements:

    ``<ngens>``
        This element is mandatory. It requires a positive non-zero integer} argument which defines the number of generations of cycles that the simulation will run.

    ``<random_seed>``
        Seed for pseudo-random number generator. The command-line specified seed (using the "``-z``" flag) overrides this, and if neither are specified, then the seed defaults to system time.

    ``<fitness_dimensions>``
        This element takes a single non-zero positive integer as its content, and specifies the number of fitness factors (trait types, dimensions of environment variables, etc.) in the simulation. If not specified, it defaults to 10.

    ``<global_selection_strength>``
        This requires a positive real number argument, and weights the overall multi-dimensional Euclidean distance between the vector of an organism's traits and the corresponding environmental optima. A value of 0.0 means that **no** selection takes place. A value of 1.0 results in the survival probability of the organisms given directly by the exponentiated distance, while higher values increase the strength of selection by lowering the survival probability of an organism for a given trait-optima distance. If not specified, defaults to 1.0.

For example::

    <system>
        <fitness_dimensions>1</fitness_dimensions>
        <global_selection_strength>1.0</global_selection_strength>
        <ngens>1000</ngens>
    </system>

The ``<lineages>`` Element: Defining the Biota
----------------------------------------------

This mandatory element defines the ecologies of all species or lineages in the simulation.
It contains one or more ``<lineage>`` child element, with each ``<lineage>`` child element defining the ecology of a single species in the simulation.

A ``<lineage>`` element in turn has one mandatory attribute, ``id``, which serves as the identifier or label for the species it defines, as well as the following child elements:

    ``fitness_trait_default_genotypes``
        The traits of organisms evolve under the specified selection pressures once the simulation begins, but the initial values for the traits for the first generation of organisms of each lineage are specified by this element, the contents of which should be a space-delimited list of numeric values.
        If not given, this defaults to all zeros.

    ``fitness_trait_relative_selection_weights``
        The contents of this element should be a space-delimited list of numeric values that provide the *relative* weights of the selection pressures for each of traits in the fitness function
        For example, "``1 1 1 1 1``" models a niche in which the organism is equally sensitive to all 5 environmental dimensions, while "``5 1 1 1 0``" models a nich in which the organism is extremely sensitive to the first environmental dimension, and, conversely extremely insensitive to the last environmental dimension, while having a moderate sensitivity to the remaining middle three dimensions.
        As these are relative weights, then "``1 1 1 1 1``" and "``2 2 2 2 2``" etc., all describe the same niche.
        If not provided, then this vector defaults to equal-weighting on all environment distances.

    ``fecundity``
        The number of offspring per mating: specified as positive integer value.

    ``movement_capacity``
        The element determines the number of "movement credits" available to an organism at the beginning of the migration phase, specified as a positive integer.
        This element has an attribute, "``distribution``", which can take on one of two values: "constant" or "poisson". If "constant", then all organisms get the same, fixed, movement capacity, equal to this element's value. If "poisson", then organisms get a random movement capacity, drawn from a Poisson distribution with a mean given by this element's value.

For example::

    <lineages>
        <lineage id="Zu">
            <fitness_trait_default_genotypes>0</fitness_trait_default_genotypes>
            <fecundity>16</fecundity>
            <movement_capacity distribution="constant">1</movement_capacity>
        </lineage>
        <lineage id="Zv">
            <fitness_trait_default_genotypes>0</fitness_trait_default_genotypes>
            <fecundity>1</fecundity>
            <movement_capacity distribution="poisson">9</movement_capacity>
        </lineage>
    </lineages>

The ``<initialization>`` Element: Defining the Starting Conditions
------------------------------------------------------------------

This element defines the initialization regime of the simulation.
In the initialization regime, the landscape is seeded with populations of organisms of various species, and the simulation is run under a particular set of environmental conditions, until all tracked neutral locii have coalesced into their respective common ancestors.
This way, the genetic structure and associated genealogies can be calibrated or conditioned under a known scenario or geo-demographic model before the main cycles of the simulation begin.
The ``<initialization>`` element has a ``<populations>`` element and an ``<environment>`` element as child elements::

    <initialization>
        <populations>
        :
        :
        </populations>
        <environment>
        :
        :
        </environment>
    </initialization>

The ``<populations>`` element describes the initial seeding of the cells of the landscape with organisms of the various species.
The ``<populations>`` element contains multiple ``<cell>`` child elements, each of which has an ``x`` and ``y`` attribute specifying the cell position on the landscape.
Each ``<cell>`` element, in turn, has multiple ``<population>`` child elements, each of which defines the number of organisms of a particular lineage to be introduced into that that cell at the beginning of the initialization phase.
For example::

    <populations>
        <cell x="0" y="0">
            <population lineage="Zu" size="100" />
            <population lineage="Zv" size="100" />
        </cell>
        <cell x="0" y="1">
            <population lineage="Zu" size="100" />
            <population lineage="Zv" size="100" />
        </cell>
        <cell x="1" y="0">
            <population lineage="Zu" size="100" />
            <population lineage="Zv" size="100" />
        </cell>
        <cell x="1" y="1">
            <population lineage="Zu" size="100" />
            <population lineage="Zv" size="100" />
        </cell>
        :
        :
    </populations>

The ``<environment>`` element describes the geographical and evolutionary template that structures or conditions the organisms in the initialization phase: the connectivity between cells, the fitness optima of the various fitness dimensions, etc.
This syntax and semantics of this element is identical to the ``<environment>`` child element of the ``<environments>`` element, described in detail below, except that the ``gen`` attribute is ignored.

A complete example of an initialization section is seen here::

    <initialization>
        <populations>
            <cell x="0" y="0">
                <population lineage="Zu" size="100" />
                <population lineage="Zv" size="100" />
            </cell>
            <cell x="0" y="1">
                <population lineage="Zu" size="100" />
                <population lineage="Zv" size="100" />
            </cell>
            <cell x="1" y="0">
                <population lineage="Zu" size="100" />
                <population lineage="Zv" size="100" />
            </cell>
            <cell x="1" y="1">
                <population lineage="Zu" size="100" />
                <population lineage="Zv" size="100" />
            </cell>
        </populations>
        <environment>
            <carrying_capacity>cc100.asc</carrying_capacity>
            <movement_costs lineage="Zu">g25x25_mv1.asc</movement_costs>
            <movement_costs lineage="Zv">g25x25_mv9.asc</movement_costs>
            <fitness_trait_optima trait="0">trait_unif0.asc</fitness_trait_optima>
            <fitness_trait_optima trait="1">trait_rand1.asc</fitness_trait_optima>
            <fitness_trait_optima trait="2">trait_rand2.asc</fitness_trait_optima>
        </environment>
    </initialization>

The ``<environments>`` Element: Defining the Landscape and Climatic History
---------------------------------------------------------------------------

The ``<environments>`` element controls the geographical template (i.e., the connectivity and movement costs of the landscape), carrying capacities and fitness regimes over the course of the simulation.
This element consists of a list of ``<Environment>`` elements, with each ``<environment>`` element encapsulating the suite of movement costs, carrying capacities, fitness trait optima, etc. to be activate at a particular generation.

The ``<environment>`` element has a single mandatory attribute, ``gen``, which takes a positive integer value specifying the generation number that this element's directive should take affect (with the first generation of the simulation being generation 0). For example::

    <environments>
        <environment gen="100"> ... </environment>
        <environment gen="200"> ... </environment>
    </environments>

The child elements of the ``<environment>`` element can be one of the following:

    ``<carrying_capacity>``
        This element specifies the carrying capacity of each cell of the landscape, i.e. the maximum number of organisms that can occupy the cell at the end of each generation. The content of this element should be an ESRI ASCII format grid file of the same dimensions as the landscape, with the cell values specifying the maximum carrying capacity of the corresponding cell on the landscape. If this element is not included for a particular ``<environment>`` suite, then the previous setting is retained.


    ``<movement_costs>``
        This element specifies the cell entry costs of the landscape for a particular lineage or species.
        It has a single mandatory attribute, ``lineage``, which should be the label or identifier of the species for which these costs should be applied.
        The content of this element should be an ESRI ASCII format grid file of the same dimensions as the landscape, with the cell values specifying the entry costs for the correspondng cells on the landscape. If this element is not included for a particular ``<environment>`` suite, then the previous setting is retained.


    ``<fitness_trait_optima>``
        This element specifies the optimum environmental value for each cell of the landscape for a particular trait.
        It has a single mandatory attribute, ``trait``, which should be the 0-based index of the trait.
        The content of this element should be an ESRI ASCII format grid file of the same dimensions as the landscape, with the cell values specifying the optimum trait value for the correspondng cells on the landscape. If this element is not included for a particular ``<environment>`` suite, then the previous setting is retained.

The following is an example of a more thoroughly specified environmental regime::

    <environments>
        <environment gen="100">
            <movement_costs lineage="Sp1">mov_sp1_gen100.asc</movement_costs>
            <movement_costs lineage="Sp2">mov_sp1_gen100.asc</movement_costs>
            <fitness_trait_optima trait="0">traits_x.asc</fitness_trait_optima>
            <fitness_trait_optima trait="1">traits_y.asc</fitness_trait_optima>
        </environment>
        <environment gen="500">
            <movement_costs lineage="Sp1">mov_sp1_gen500.asc</movement_costs>
            <carrying_capacity>cc_gen500.asc</carrying_capacity>
        </environment>
        <environment gen="1000">
            <carrying_capacity>cc_gen1000.asc</carrying_capacity>
            <movement_costs lineage="Sp2">mov_sp2_gen1000.asc</movement_costs>
            <fitness_trait_optima trait="0">traits_a.asc</fitness_trait_optima>
            <fitness_trait_optima trait="1">traits_b.asc</fitness_trait_optima>
        </environment>
    </environments>

In generation 100 of the simulation, cell entry costs are set for species "Sp1" and "Sp2", as well as the environmental fitness optima for the first and second fitness factors. Until these settings were changed in generation "100", they would have retained the values specified in the initialization phase. In addition, settings not specified in this generation (for example, the carrying capacity) would also retain the values given in the initialization phase. In generation 500, the movement costs for "Sp1" and the overall carrying capacity of landscape are changed. Again, all other settings (e.g., the movement costs for "Sp1", the fitness trait optima) remain unchanged from the values they had in generation "100".

The ``<samples>`` Element: Defining the Sampling Regime
-------------------------------------------------------

The ``<samples>`` element serves as container for one or more ``<sample>`` elements, each of which describes the sampling design for a set of genealogies and occurrence data that will be sampled at a particular point in time during the simulation for a particular lineage.
Each ``<sample>`` element has two mandatory attributes: the ``gen`` attribute is a positive integer that defines the 0-based index of the generation number that this sample should be taken, and the ``lineage`` attribute which is a string giving the identifier of the lineage to be sampled::

    <samples>
        <sample gen="100" lineage="Sp1"> ... </sample>
        <sample gen="100" lineage="Sp2"> ... </sample>
        <sample gen="200" lineage="Sp1"> ... </sample>
        :
    </samples>

Each sample, by default, will result in three files:

    * an occurrence matrix, showing the number of individuals of the specified lineage in each cell of the landscape
    * a genealogy showing the relationship of the alleles of the neutral haploid locus of every organism of the specified lineage from every cell of the landscape

ChangeLog
2009-03-06  Jeet Sukumaran (jeetsukumaran@gmail.com)
    
    * Version 0.0.4.

    * Multiple diploid locus (+ single haploid locus) tracking.

2009-03-05  Jeet Sukumaran (jeetsukumaran@gmail.com)

    * Version 0.0.3.

    * Diploid locus tracking.

2009-03-06  Jeet Sukumaran (jeetsukumaran@gmail.com)

    * Version 0.0.2.

    * Haploid locus tracking.

2009-03-06  Jeet Sukumaran (jeetsukumaran@gmail.com)

    * Version 0.0.1.

    * Biparental organism pedigree tracking.

2009-03-06  Jeet Sukumaran (jeetsukumaran@gmail.com)

    * Version 0.0.0.

    * Prototype.


ginkgopy/ginkgo/config.py
#! /usr/bin/env python

###############################################################################
##
## GINKGO Biogeographical Evolution Simulator Post-Processing Library.
##
## Copyright 2009 Jeet Sukumaran and Mark T. Holder.
##
## This program is free software; you can redistribute it and#or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 3 of the License, or
## (at your option) any later version.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License along
## with this program. If not, see <http:##www.gnu.org#licenses#>.
##
###############################################################################

"""
Helps in building configuration files.
"""

from cStringIO import StringIO

INDENT_SIZE = 4

class GinkgoConfiguration(object):

    def __init__(self, **kwargs):
        self.title = kwargs.get("title", "results")
        self.log_frequency = kwargs.get("log_frequency", None)
        self.multifurcating_trees = kwargs.get("multifurcating_trees", None)
        self.final_output = kwargs.get("final_output", None)
        self.full_complement_diploid_trees = kwargs.get("full_complement_diploid_trees", None)
        self.system = kwargs.get("system", None)
        self.landscape = kwargs.get("landscape", None)
        self.lineages = kwargs.get("lineages", [])
        self.initialization_regime = kwargs.get("initialization_regime", None)
        self.environments = kwargs.get("environments", [])
        self.samples = kwargs.get("samples", [])

    def __str__(self):
        s = StringIO()
        s.write('<?xml version="1.0"?>\n')
        s.write('<ginkgo')
        indent = INDENT_SIZE * ' '
        if self.title is not None:
            s.write('\n%stitle="%s"' % (indent, self.title))
        if self.log_frequency is not None:
            s.write('\n%slog_frequency="%s"' % (indent, self.log_frequency))
        if self.multifurcating_trees is not None:
            s.write('\n%smultifurcating_trees="%s"' % (indent, self.multifurcating_trees))
        if self.final_output is not None:
            s.write('\n%sfinal_output="%s"' % (indent, self.final_output))
        if self.full_complement_diploid_trees is not None:
            s.write('\n%sfull_complement_diploid_trees="%s"' % (indent, self.full_complement_diploid_trees))
        s.write('>\n')
        sub_indent = indent * 1
        if self.system is not None:
            s.write(str(self.system))
        if self.landscape is not None:
            s.write(str(self.landscape))
        if self.lineages:
            s.write('%s<lineages>\n' % sub_indent)
            for lineage in self.lineages:
                s.write(str(lineage))
            s.write('%s</lineages>\n' % sub_indent)
        if self.initialization_regime is not None:
            s.write(str(self.initialization_regime))
        if self.environments:
            s.write('%s<environments>\n' % sub_indent)
            for environment in self.environments:
                s.write(str(environment))
            s.write('%s</environments>\n' % sub_indent)
        if self.samples:
            s.write('%s<samples>\n' % sub_indent)
            for sample in self.samples:
                s.write(str(sample))
            s.write('%s</samples>\n' % sub_indent)
        s.write('</ginkgo>\n')
        return s.getvalue()

class System(object):

    def __init__(self, **kwargs):
        self.random_seed = kwargs.get("random_seed", None)
        self.fitness_dimensions = kwargs.get("fitness_dimensions", None)
        self.global_selection_strength = kwargs.get("global_selection_strength", None)
        self.ngens = kwargs.get("ngens", None)

    def __str__(self):
        s = StringIO()
        indent = (INDENT_SIZE * ' ')
        sub_indent = indent * 2
        s.write('%s<system>' % indent)
        if self.random_seed is not None:
            s.write('\n%s<random_seed>%s</random_seed>' % (sub_indent, self.random_seed))
        if self.fitness_dimensions is not None:
            s.write('\n%s<fitness_dimensions>%s</fitness_dimensions>' % (sub_indent, self.fitness_dimensions))
        if self.global_selection_strength is not None:
            s.write('\n%s<global_selection_strength>%s</global_selection_strength>' % (sub_indent, self.global_selection_strength))
        if self.ngens is not None:
            s.write('\n%s<ngens>%s</ngens>' % (sub_indent, self.ngens))
        s.write('\n%s</system>\n' % indent)
        return s.getvalue()

class Landscape(object):

    def __init__(self, ncols, nrows, **kwargs):
        self.ncols = ncols
        self.nrows = nrows
        self.default_cell_carrying_capacity = kwargs.get("default_cell_carrying_capacity", None)

    def __str__(self):
        s = StringIO()
        indent = (INDENT_SIZE * ' ')
        sub_indent = indent * 2
        s.write('%s<landscape ncols="%s" nrows="%s">' % (indent, self.ncols, self.nrows))
        if self.default_cell_carrying_capacity is not None:
            s.write('\n%s<default_cell_carrying_capacity>%s</default_cell_carrying_capacity>' % (sub_indent, self.default_cell_carrying_capacity))
        s.write('\n%s</landscape>\n' % indent)
        return s.getvalue()

class Lineage(object):

    def __init__(self, lineage_id, **kwargs):
        self.lineage_id = lineage_id
        self.fitness_trait_relative_selection_weights = kwargs.get("fitness_trait_relative_selection_weights", None)
        self.fitness_trait_default_genotypes = kwargs.get("fitness_trait_default_genotypes", None)
        self.fecundity = kwargs.get("fecundity", None)
        self.__movement_capacity = None
        self.movement_capacity = kwargs.get("movement_capacity", None)
        self.indent_level = kwargs.get('indent_level', 2)

    def _set_movement_capacity(self, val):
        if not (isinstance(val, list) or isinstance(val, tuple))\
                or len(val) != 2 \
                or not isinstance(val[0], str):
            raise ValueError("Expecting tuple of (<dist>, <val>) for movement capacity, " \
                           + "where <dist> is string specifying the distribution ('constant' or 'poisson'), " \
                           + "and <val> is either the fixed movement capacity value or the mean of the Poisson")
        self.__movement_capacity = val

    def _get_movement_capacity(self):
        return self.__movement_capacity

    movement_capacity = property(_get_movement_capacity, _set_movement_capacity)

    def __str__(self):
        top_indent = (self.indent_level * INDENT_SIZE) * ' '
        s = StringIO()
        s.write('%s<lineage id="%s">\n' % (top_indent, self.lineage_id))
        sub_indent = ((self.indent_level + 1) * INDENT_SIZE) * ' '
        if self.fitness_trait_relative_selection_weights:
            s.write('%s<fitness_trait_relative_selection_weights>%s</fitness_trait_relative_selection_weights>\n' %
                    (sub_indent, (" ".join([str(x) for x in self.fitness_trait_relative_selection_weights]))))
        if self.fitness_trait_default_genotypes:
            s.write('%s<fitness_trait_default_genotypes>%s</fitness_trait_default_genotypes>\n' %
                    (sub_indent, (" ".join([str(x) for x in self.fitness_trait_default_genotypes]))))
        if self.fecundity:
            s.write('%s<fecundity>%s</fecundity>\n' % (sub_indent, self.fecundity))
        if self.movement_capacity:
            s.write('%s<movement_capacity distribution="%s">%s</movement_capacity>\n' \
                    % (sub_indent, self.movement_capacity[0], self.movement_capacity[1]))
        s.write('%s</lineage>\n' % (top_indent))
        return s.getvalue()

class EnvironmentGrid(object):

    def __init__(self, type_name, attribute_dict, grid_filepath):
        self.type_name = type_name
        self.attribute_dict = attribute_dict
        self.grid_filepath = grid_filepath

    def __str__(self):
        s = StringIO()
        s.write("<%s" % self.type_name)
        if self.attribute_dict:
            parts = []
            for k, v in self.attribute_dict.items():
                parts.append('%s="%s"' % (k,v))
            s.write(" %s" % (" ".join(parts)))
        s.write(">%s</%s>" % (self.grid_filepath, self.type_name))
        return s.getvalue()

class CarryingCapacityGrid(EnvironmentGrid):

    def __init__(self, grid_filepath):
        EnvironmentGrid.__init__(self,
                "carrying_capacity",
                attribute_dict={},
                grid_filepath=grid_filepath)

#class MovementProbabilitiesGrid(EnvironmentGrid):
#
#    def __init__(self, lineage, grid_filepath):
#        EnvironmentGrid.__init__(self,
#                "movement_probabilities",
#                attribute_dict={'lineage': lineage},
#                grid_filepath=grid_filepath)

class MovementCostsGrid(EnvironmentGrid):

    def __init__(self, lineage, grid_filepath):
        EnvironmentGrid.__init__(self,
                "movement_costs",
                attribute_dict={'lineage': lineage},
                grid_filepath=grid_filepath)

class FitnessTraitOptimaGrid(EnvironmentGrid):

    def __init__(self, trait, grid_filepath):
        EnvironmentGrid.__init__(self,
                "fitness_trait_optima",
                attribute_dict={'trait': trait},
                grid_filepath=grid_filepath)

class Environment(object):

    def __init__(self, gen, indent_level=2):
        self.gen = gen
        self.grids = []
        self.indent_level = indent_level

    def __str__(self):
        if len(self.grids) == 0:
            return ""
        parts = []
        top_indent = (self.indent_level * INDENT_SIZE) * ' '
        if self.gen is not None:
            gen = ' gen="%d"' % self.gen
        else:
            gen = ""
        parts.append('%s<environment%s>' % (top_indent, gen))
        sub_indent = ((self.indent_level + 1) * INDENT_SIZE) * ' '
        for grid in self.grids:
            parts.append('%s%s' % (sub_indent, str(grid)))
        parts.append('%s</environment>\n' % top_indent)
        return "\n".join(parts)

class InitializationCellPopulations(object):

    def __init__(self, **kwargs):
        self.x = kwargs.get("x", None)
        self.y = kwargs.get("y", None)
        self.index = kwargs.get("index", None)
        self.pops = kwargs.get("pops", {})
        self.indent_level = kwargs.get("indent_level", 3)

    def __str__(self):
        parts = []
        top_indent = (self.indent_level * INDENT_SIZE) * ' '
        if self.index is not None:
            parts.append('%s<cell index="%s">' % (top_indent, self.index))
        elif self.x is not None and self.y is not None:
            parts.append('%s<cell x="%s" y="%s">' % (top_indent, self.x, self.y))
        else:
            raise TypeError("Either 'index' or both 'x' and 'y' must be specified")
        sub_indent = ((self.indent_level+1) * INDENT_SIZE) * ' '
        for sp, sz in self.pops.items():
            parts.append('%s<population lineage="%s" size="%s" />' % (sub_indent, sp, sz))
        parts.append("%s</cell>" % top_indent)
        return "\n".join(parts)

class InitializationRegime(object):

    def __init__(self, pops, environment, max_cycles=None, indent_level=1):
        self.pops = pops
        self.environment = environment
        self.max_cycles = max_cycles
        self.indent_level = indent_level

    def __str__(self):
        parts = []
        top_indent = (self.indent_level * INDENT_SIZE) * ' '
        sub_indent = ((self.indent_level+1) * INDENT_SIZE) * ' '
        if self.max_cycles is not None:
            max_cycles = ' max_cycles="%s"' % self.max_cycles
        else:
            max_cycles = ""
        parts.append('%s<initialization%s>' % (top_indent, max_cycles))
        e = str(self.environment)
        if e.endswith('\n'):
            e = e[:-1]
        parts.append(e)
        parts.append('%s<populations>' % (sub_indent))
        for cp in self.pops:
            parts.append(str(cp))
        parts.append('%s</populations>' % (sub_indent))
        parts.append("%s</initialization>\n" % top_indent)
        return "\n".join(parts)

class Sample(object):

    def __init__(self, lineage_id, gen, **kwargs):
        self.lineage_id = lineage_id
        self.gen = gen
        self.label = kwargs.get('label', None)
        self.individuals_per_cell = kwargs.get('individuals_per_cell', None)
        self.cell_coordinates = kwargs.get('cell_coordinates', None)
        self.cell_indexes = kwargs.get('cell_indexes', None)
        self.indent_level = kwargs.get('indent_level', 2)
        self.build_trees = kwargs.get('build_trees', True)

    def __str__(self):
        top_indent = (self.indent_level * INDENT_SIZE) * ' '
        if self.label:
            label = ' label="%s"' % self.label
        else:
            label = ''
        parts = []
        if self.build_trees:
            close = '>'
        else:
            close = ' trees="False" />\n'
        parts.append('%s<sample gen="%d" lineage="%s"%s%s' % (top_indent, self.gen, self.lineage_id, label, close))
        if self.build_trees:
            sub_indent = ((self.indent_level + 1) * INDENT_SIZE) * ' '
            if self.individuals_per_cell:
                parts.append('%s<individuals_per_cell>%s</individuals_per_cell>' % (sub_indent, self.individuals_per_cell))
            if self.cell_indexes or self.cell_coordinates:
                parts.append('%s<cells>' % sub_indent)
                sub_sub_indent = ((self.indent_level + 2) * INDENT_SIZE) * ' '
                if self.cell_indexes:
                    for c in self.indexes:
                        parts.append('%s<cell index="%s" />' % (sub_sub_indent, c))
                if self.cell_coordinates:
                    for c in self.cell_coordinates:
                        parts.append('%s<cell x="%s" y="%s"/>' % (sub_sub_indent, c[0], c[1]))
                parts.append('%s<cells>' % sub_indent)
            parts.append('%s</sample>\n' % top_indent)
        return "\n".join(parts)

if __name__ == "__main__":
    num_fitness_traits = 1

    g = GinkgoConfiguration(
            title="results1",
            log_frequency=100,
            multifurcating_trees=True,
            final_output=False,
            full_complement_diploid_trees=False,
            system=System(random_seed=3124,
                         fitness_dimensions=num_fitness_traits,
                         global_selection_strength=1.0,
                         ngens=1000),
            landscape=Landscape(ncols=10, nrows=10, default_cell_carrying_capacity=100)
        )
    cells = [(1,1),(2,1),(3,1),(4,1),
             (1,2),(2,2),(3,2),(4,2)]
    s1 = Lineage("Zx",
            fitness_trait_relative_selection_weights=[1]*num_fitness_traits,
            fitness_trait_default_genotypes=[0]*num_fitness_traits,
            fecundity=16,
            movement_capacity=("poisson", 10))
    g.lineages.append(s1)

    e0 = Environment(None)
    e0.grids.append(CarryingCapacityGrid('cc.grd'))
    e0.grids.append(MovementCostsGrid('Zx', 'movp_zx.grd'))
    e0.grids.append(FitnessTraitOptimaGrid('1', 'movp_zx.grd'))
    cell_pops = []
    for cell in cells:
        cell_pops.append(InitializationCellPopulations(x=cell[0], y=cell[1], pops={"Zx": 100}))
    init_reg = InitializationRegime(pops=cell_pops, environment=e0)
    g.initialization_regime = init_reg

    e1 = Environment(1)
    e1.grids.append(CarryingCapacityGrid('cc.grd'))
    e1.grids.append(MovementCostsGrid('Zx', 'movp_zx.grd'))
    e1.grids.append(FitnessTraitOptimaGrid('1', 'movp_zx.grd'))
    g.environments.append(e1)

    for t in xrange(10):
        gen = (t + 10) * 1000
        s = Sample(lineage_id="Zx", gen=gen, individuals_per_cell=10, cell_coordinates=cells)
        g.samples.append(s)

    print(str(g))
